feat(errors) - Support Error upsampling on Issue Alerts (current system)

yuvmen opened the pull request:
- EventFrequency and EventPercentage Issue Alerts now correctly handle counting error on error upsampling enabled projects.
- Because "count()" aggregation was assumed inside SnubaTSDB class, was necessary to pass project_ids into TSDB to make decision there.

Context:
Error upsampling which is enabled for specific customers means we need to count errors differently. This PR changes the way Issue Alerts count errors under the hood for specific projects (using an option). The "upsampled_count" aggregation internal implementation is something that was already done in previous commits, here we just swap what aggregation we are doing - "upsampled_count" instead of "count".

I realize these changes are on the current system and should also be added to the new workflow engine, this will be in a follow up PR.

yuvmen committed:
feat(errors) - Support Error upsampling on Issue Alerts

- EventFrequency and EventPercentage Issue Alerts now correctly handle counting
error on error upsampling enabled projects.
- Because "count()" aggergation was assumed inside SnubaTSDB class, was necessary to pass
project_ids into TSDB to make decision there.

yuvmen committed:
correct project_ids being passed only to tsdb functions which support it

yuvmen committed:
corrected addition of project_ids to get_snuba_query_result and some test refactors

ceorourke reviewed (approved):
Change lgtm - might want to run it by @cathteng just in case

cathteng commented on src/sentry/tsdb/snuba.py:
```diff
@@ -768,16 +798,26 @@ def get_range(
-            aggregation=aggregation_override or self.get_aggregate_function(model),
-            group_on_time=True,
+            aggregation=aggregation,
+            group_on_time=group_on_time,
```
Should this be `true`? @ceorourke ?

yuvmen replied:
oh damn I missed this, good job you and cursor 👍  will fix

cathteng commented on src/sentry/tsdb/snuba.py:
```diff
@@ -753,6 +763,26 @@ def get_sums_data(
+            return False
+
+        try:
+            return are_any_projects_error_upsampled(list(project_ids))
```
I'd expect this to be a single `project_id` since delayed processing is batched per project 👍

yuvmen replied:
I thought so at first, but there is the case of `batch_query_hook` usecase with the call to get_chunked_result, where it gives like a list of groups and then I will be getting a list of project Ids for real right?

cathteng replied:
The groups should be in the same project, it's called here and each task is for 1 project
https://github.com/getsentry/sentry/blob/636371fe3e34c3ee8917be3ba7346329e59c7d0c/src/sentry/rules/processing/delayed_processing.py#L410-L417

yuvmen replied:
hmm I see.. I wonder if I should be strict about it in code, or just allow it like it is, right now if that ever changes this will still work

cathteng commented on src/sentry/rules/conditions/event_frequency.py:
```diff
@@ -726,6 +743,7 @@ def get_snuba_query_result(
         environment_id: int,
         referrer_suffix: str,
         group_on_time: bool = False,
+        project_ids: list[int] | None = None,
```
We are only upsampling for error count, not user count?

cathteng reviewed (commented):
I think this looks good but I have some outstanding questions. How do you expect to revert if this is throwing errors? Turn upsampling off for the specific projects?

yuvmen committed:
fix get_data call in get_range to always set grop_on_time to true

yuvmen commented:
@cathteng yea we have a backend option to control it, though I am touching some very common internals I worry more about not breaking the normal use case, breaking the upsampled_projects is not a big deal we are still in integration there over one demo project in US.
I tried to test as much as I could, I will be monitoring to see if goes over smooth.

yuvmen committed:
fix typing issue - retrigger CI

yuvmen committed:
Merge branch 'master' into yuvmen/error-upsampling-issue-alerts

yuvmen committed:
Merge branch 'master' into yuvmen/error-upsampling-issue-alerts

yuvmen committed:
fix typing issue on get_range missing aggregation_override

this was an addition I made in another PR for event-stats, it should actually be refactored
now that there is a better way to do it, but I will save it for a follow up PR and not here
